src/hotspot/share/opto/loopopts.cpp

```diff
@@ -2581,11 +2721,18 @@ void PhaseIdealLoop::clone_loop( IdealLoopTree *loop, Node_List &old_new, int dd
   Node_List *split_cex_set = nullptr;
   fix_data_uses(loop->_body, loop, mode, outer_loop, new_counter, old_new, worklist, split_if_set, split_bool_set, split_cex_set);
 
-  for (uint i = 0; i < extra_data_nodes.size(); i++) {
-    Node* old = extra_data_nodes.at(i);
-    clone_loop_handle_data_uses(old, old_new, loop, outer_loop, split_if_set,
-                                split_bool_set, split_cex_set, worklist, new_counter,
-                                mode);
+  if (mode != InsertAtomicPost) {
+    for (uint i = 0; i < extra_data_nodes.size(); i++) {
+      Node* old = extra_data_nodes.at(i);
+      clone_loop_handle_data_uses(old, old_new, loop, outer_loop, split_if_set,
+                                  split_bool_set, split_cex_set, worklist, new_counter,
+                                  mode);
+    }
+  } else {
+    for (uint i = 0; i < extra_data_nodes.size(); i++) {
+      Node* old = extra_data_nodes.at(i);
+      handle_data_uses_for_atomic_post_loop(old, old_new, loop, outer_loop, worklist, new_counter);
+    }
```

Comment:
Probably we will never add any new cases, but if so, it might be better to have the "default" case last. We might have a switch, or some if-elseif...